src/Models/IoTWorkspaceProject.ts

```diff
@@ -359,35 +346,38 @@ export class IoTWorkspaceProject extends IoTWorkbenchProjectBase {
   }
 
   /**
-   * Init Azure components and azure configs.
+   * Load project config and init Azure components.
    * @param scaffoldType scaffold type
    */
-  private async initAzureConfig(scaffoldType: ScaffoldType): Promise<void> {
-    this.validateProjectRootPath("initialize azure configurations", scaffoldType);
-
+  private async loadAzureConfigAndComponents(scaffoldType: ScaffoldType): Promise<void> {
     const azureConfigFileHandler = new azureComponentConfigModule.AzureConfigFileHandler(this.projectRootPath);
     await azureConfigFileHandler.createIfNotExists(scaffoldType);
 
     const componentConfigs = await azureConfigFileHandler.getSortedComponents(scaffoldType);
 
     if (componentConfigs.length === 0) {
       // Support backward compact
-      await this.initAzureComponentsWithoutConfig(scaffoldType);
+      this.initAzureComponentsWithoutConfig(scaffoldType);
     } else {
-      await this.initAzureComponentsWithConfig(scaffoldType, componentConfigs);
+      await this.loadAzureComponentsByConfig(componentConfigs);
     }
   }
 
-  private async createAzureComponentsWithProjectType(
-    projectType: ProjectTemplateType,
-    scaffoldType: ScaffoldType,
-    workspaceConfig: Workspace
-  ): Promise<void> {
-    this.validateProjectRootPath("init azure components", scaffoldType);
-
-    // initialize the storage for azure component settings
-    const azureConfigFileHandler = new azureComponentConfigModule.AzureConfigFileHandler(this.projectRootPath);
-    await azureConfigFileHandler.createIfNotExists(scaffoldType);
+  /**
+   * Init azure components based on project template type.
+   * Update component list with azure components.
+   * Update azure components information in workspace configuration.
+   * @param projectType project template type
+   * @param workspaceConfig workspace configuration
+   */
+  private initAzureComponents(projectType: ProjectTemplateType, workspaceConfig: Workspace): void {
+    if (!this.projectRootPath) {
+      throw new ArgumentEmptyOrNullError(
+        "init azure components",
+        `project root path: ${this.projectRootPath}`,
+        "Please initialize the project first."

```

Comment:
I would prefer formatting the error message and then pass it to the `ArgumentEmptyOrNullError` rather than add complexity to the constructor. This makes the code less readable.

Reply:
Will discuss further and fix in new PR for all error types.